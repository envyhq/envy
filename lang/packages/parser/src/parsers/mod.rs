mod attribute_block_parser;
mod module_declaration_parser;
mod provider_declaration_parser;
pub mod source_file_parser;
mod var_block_parser;
mod var_declaration_parser;
