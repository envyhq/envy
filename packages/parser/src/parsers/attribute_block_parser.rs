use crate::{
    abstract_syntax_tree::{AbstractSyntaxNode, DeclarationNode},
    attributes::{AttributeDeclarationNode, PartialAttributeDeclarationNode},
    Parser,
};
use nv_lexer::{tokens::LexerSymbol, LexerToken};

pub struct AttributeBlockParser {
    pub ast_block: Vec<AbstractSyntaxNode>,
    pub tokens: Vec<LexerToken>,

    buffer: Vec<LexerToken>,
}

impl Parser for AttributeBlockParser {
    fn parse(&mut self) -> usize {
        let bound_tokens = self.tokens.clone();
        let mut tokens = bound_tokens.iter().enumerate();

        self.buffer.clear();

        let mut processed_count = 0;

        let mut partial_declaration = PartialAttributeDeclarationNode {
            identifier: None,
            value: None,
        };

        while let Some((_index, token)) = tokens.next() {
            let token = token.to_owned();

            processed_count += 1;

            match token {
                LexerToken::Identifier(identifier) => {
                    partial_declaration.identifier = Some(identifier);
                }
                LexerToken::Literal(value) => {
                    partial_declaration.value = Some(value);
                }
                LexerToken::Symbol(LexerSymbol::BlockCloseCurly) => {
                    break;
                }
                _ => {
                    self.buffer.push(token);
                }
            };

            let declaration: Result<AttributeDeclarationNode, _> =
                partial_declaration.clone().try_into();

            if declaration.is_ok() {
                self.ast_block.push(AbstractSyntaxNode::Declaration(
                    DeclarationNode::AttributeDeclaration(declaration.unwrap().clone()),
                ));
                partial_declaration = PartialAttributeDeclarationNode {
                    identifier: None,
                    value: None,
                };

                continue;
            }
        }

        processed_count
    }
}

impl AttributeBlockParser {
    pub fn new(tokens: Vec<LexerToken>) -> Self {
        Self {
            ast_block: vec![],
            tokens,
            buffer: vec![],
        }
    }
}
