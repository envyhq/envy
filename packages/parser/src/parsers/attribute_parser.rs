use crate::Parser;

pub struct AttributeBlockParser {}

impl Parser for AttributeBlockParser {
    fn parse(&mut self) -> usize {
        0
    }
}

impl AttributeBlockParser {
    pub fn new() -> Self {
        Self {}
    }
}
