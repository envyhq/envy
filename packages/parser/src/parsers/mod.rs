mod attribute_parser;
mod module_declaration_parser;
mod provider_declaration_parser;
pub mod source_file_parser;
mod var_declaration_parser;
