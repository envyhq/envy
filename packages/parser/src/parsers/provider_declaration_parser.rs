use crate::{
    abstract_syntax_tree::{AbstractSyntaxNode, DeclarationNode},
    parsers::attribute_block_parser::AttributeBlockParser,
    providers::{PartialProviderDeclarationNode, ProviderDeclarationNode},
    Parser,
};
use nv_lexer::{tokens::LexerSymbol, LexerTokenKind};

pub struct ProviderDeclarationParser {
    pub ast_fragment: Option<AbstractSyntaxNode>,
    pub tokens: Vec<LexerTokenKind>,

    buffer: Vec<LexerTokenKind>,
}

impl Parser for ProviderDeclarationParser {
    fn parse(&mut self) -> usize {
        let bound_tokens = self.tokens.clone();
        let mut tokens = bound_tokens.iter().enumerate();

        self.buffer.clear();

        let mut processed_count = 0;

        let mut partial_declaration = PartialProviderDeclarationNode {
            identifier: None,
            type_value: None,
            attributes: vec![],
        };

        while let Some((index, token)) = tokens.next() {
            let token = token.to_owned();

            processed_count += 1;

            let sub_tokens = &bound_tokens[index..].to_vec();
            let sub_tokens = sub_tokens.to_vec();

            match token {
                LexerTokenKind::Identifier(identifier) => {
                    partial_declaration.identifier = Some(identifier);

                    continue;
                }
                LexerTokenKind::ProviderType(type_value) => {
                    partial_declaration.type_value = Some(type_value);

                    continue;
                }
                LexerTokenKind::Symbol(LexerSymbol::BlockOpenCurly) => {
                    let mut parser = AttributeBlockParser::new(sub_tokens.clone());
                    // -1 because we dont want to double count the block open curly
                    let count = parser.parse() - 1;

                    partial_declaration.attributes = parser
                        .ast_block
                        .iter()
                        .map(|declaration| declaration.clone().into())
                        .collect();

                    processed_count += count;
                    if count > 0 {
                        tokens.nth(count - 1);
                    }

                    let declaration: Result<ProviderDeclarationNode, _> =
                        partial_declaration.clone().try_into();

                    if declaration.is_ok() {
                        self.ast_fragment = Some(AbstractSyntaxNode::Declaration(
                            DeclarationNode::ProviderDeclaration(declaration.unwrap().clone()),
                        ));
                        return processed_count;
                    }

                    continue;
                }
                LexerTokenKind::Symbol(LexerSymbol::Newline) => {
                    let declaration: Result<ProviderDeclarationNode, _> =
                        partial_declaration.clone().try_into();

                    if declaration.is_ok() {
                        self.ast_fragment = Some(AbstractSyntaxNode::Declaration(
                            DeclarationNode::ProviderDeclaration(declaration.unwrap().clone()),
                        ));

                        return processed_count;
                    }

                    continue;
                }
                _ => {
                    self.buffer.push(token);
                    continue;
                }
            };
        }

        let declaration: Result<ProviderDeclarationNode, _> =
            partial_declaration.clone().try_into();

        if declaration.is_ok() {
            self.ast_fragment = Some(AbstractSyntaxNode::Declaration(
                DeclarationNode::ProviderDeclaration(declaration.unwrap().clone()),
            ));

            return processed_count;
        }

        processed_count
    }
}

impl ProviderDeclarationParser {
    pub fn new(tokens: Vec<LexerTokenKind>) -> Self {
        Self {
            ast_fragment: None,
            tokens,
            buffer: vec![],
        }
    }
}
