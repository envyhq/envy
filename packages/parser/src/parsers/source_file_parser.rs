use super::{
    module_declaration_parser::ModuleDeclarationParser,
    provider_declaration_parser::ProviderDeclarationParser,
};
use crate::{
    abstract_syntax_tree::{AbstractSyntaxNode, SourceFileNode},
    parser::ParserResult,
    parsers::var_declaration_parser::VarDeclarationParser,
    Parser,
};
use nv_lexer::{
    tokens::{LexerDeclarationKeyword, LexerToken},
    LexerKeyword, LexerTokenKind, LexerVarModifierKeyword,
};
use std::sync::{Arc, Mutex};

#[derive(Debug)]
pub struct SourceFileParser;

impl SourceFileParser {
    pub fn parse(tokens: &[LexerToken]) -> (usize, Arc<AbstractSyntaxNode>) {
        let root = Arc::new(AbstractSyntaxNode::SourceFile(Arc::new(SourceFileNode {
            declarations: Mutex::new(vec![]),
        })));
        let mut tokens_iter = tokens.iter().enumerate();

        let mut processed_count = 0;

        while let Some((index, token)) = tokens_iter.next() {
            processed_count += 1;
            let sub_tokens = &tokens[index..].to_vec();
            let sub_tokens = sub_tokens.to_vec();

            let result = match token.kind.clone() {
                LexerTokenKind::Keyword(LexerKeyword::VarModifierKeyword(
                    LexerVarModifierKeyword::Pub,
                ))
                | LexerTokenKind::Keyword(LexerKeyword::DeclarationKeyword(
                    LexerDeclarationKeyword::Var,
                )) => {
                    let (count, parsed_fragment) =
                        VarDeclarationParser::parse(&sub_tokens, Arc::downgrade(&root));

                    ParserResult {
                        processed_count: count - 1,
                        ast_fragment: parsed_fragment,
                    }
                }
                LexerTokenKind::Keyword(LexerKeyword::DeclarationKeyword(keyword)) => match keyword
                {
                    LexerDeclarationKeyword::Provider => {
                        let (count, parsed_fragment) =
                            ProviderDeclarationParser::parse(&sub_tokens, Arc::downgrade(&root));

                        ParserResult {
                            processed_count: count - 1,
                            ast_fragment: parsed_fragment,
                        }
                    }
                    LexerDeclarationKeyword::Module => {
                        let (count, parsed_fragment) =
                            ModuleDeclarationParser::parse(&sub_tokens, Arc::downgrade(&root));

                        ParserResult {
                            processed_count: count - 1,
                            ast_fragment: parsed_fragment,
                        }
                    }
                    _ => {
                        continue;
                    }
                },
                _ => {
                    continue;
                }
            };

            if let Some(AbstractSyntaxNode::Declaration(declaration)) = result.ast_fragment {
                if let AbstractSyntaxNode::SourceFile(source_node) = root.as_ref() {
                    source_node
                        .declarations
                        .lock()
                        .unwrap()
                        .push(Arc::new(declaration));
                }
            }

            processed_count += result.processed_count;
            if result.processed_count > 0 {
                tokens_iter.nth(result.processed_count - 1);
            }
        }

        (processed_count, root)
    }
}
