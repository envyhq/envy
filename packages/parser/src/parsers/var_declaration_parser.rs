use crate::{
    abstract_syntax_tree::{AbstractSyntaxNode, DeclarationNode, Leaf},
    parsers::attribute_block_parser::AttributeBlockParser,
    vars::{PartialVarDeclarationNode, VarDeclarationNode},
    Parser,
};
use nv_lexer::{
    tokens::{LexerSymbol, LexerToken},
    LexerKeyword, LexerTokenKind,
};
use std::sync::Weak;

pub struct VarDeclarationParser;

impl Parser<Option<AbstractSyntaxNode>> for VarDeclarationParser {
    fn parse(
        tokens: &Vec<LexerToken>,
        parent: Weak<AbstractSyntaxNode>,
    ) -> (usize, Option<AbstractSyntaxNode>) {
        let mut buffer = vec![];
        let mut tokens_iter = tokens.iter().enumerate();

        let mut ast_fragment = None;

        let mut processed_count = 0;

        let mut partial_declaration = PartialVarDeclarationNode {
            parent: parent.clone(),
            identifier: None,
            type_value: None,
            modifier: None,
            attributes: vec![],
        };

        while let Some((index, token)) = tokens_iter.next() {
            let token = token.to_owned();

            processed_count += 1;

            let sub_tokens = &tokens[index..].to_vec();
            let sub_tokens = sub_tokens.to_vec();

            match token.kind {
                LexerTokenKind::Identifier(identifier) => {
                    partial_declaration.identifier = Some(Leaf::new(identifier, token.range));

                    continue;
                }
                LexerTokenKind::Type(type_value) => {
                    partial_declaration.type_value = Some(Leaf::new(type_value, token.range));

                    continue;
                }
                LexerTokenKind::Keyword(LexerKeyword::VarModifierKeyword(modifier)) => {
                    partial_declaration.modifier = Some(Leaf::new(modifier, token.range));

                    continue;
                }
                LexerTokenKind::Symbol(LexerSymbol::BlockOpenCurly) => {
                    let (count, parsed_block) =
                        AttributeBlockParser::parse(&sub_tokens, parent.clone());

                    // -1 because we dont want to double count the block open curly
                    let count = count - 1;

                    partial_declaration.attributes = parsed_block
                        .iter()
                        .map(|attribute| attribute.clone().into())
                        .collect();

                    processed_count += count;
                    if count > 0 {
                        tokens_iter.nth(count - 1);
                    }

                    let declaration: Result<VarDeclarationNode, _> =
                        partial_declaration.clone().try_into();

                    if declaration.is_ok() {
                        ast_fragment = Some(AbstractSyntaxNode::Declaration(
                            DeclarationNode::VarDeclaration(declaration.unwrap()),
                        ));
                    }
                }
                LexerTokenKind::Symbol(LexerSymbol::Newline) => {
                    let declaration: Result<VarDeclarationNode, _> =
                        partial_declaration.clone().try_into();

                    if declaration.is_ok() {
                        ast_fragment = Some(AbstractSyntaxNode::Declaration(
                            DeclarationNode::VarDeclaration(declaration.unwrap()),
                        ));
                    }
                }
                _ => {
                    buffer.push(token);
                    continue;
                }
            };

            if ast_fragment.is_some() {
                return (processed_count, ast_fragment);
            }
        }

        let declaration: Result<VarDeclarationNode, _> = partial_declaration.clone().try_into();

        if declaration.is_ok() {
            ast_fragment = Some(AbstractSyntaxNode::Declaration(
                DeclarationNode::VarDeclaration(declaration.unwrap()),
            ));

            return (processed_count, ast_fragment);
        }

        (processed_count, ast_fragment)
    }
}
