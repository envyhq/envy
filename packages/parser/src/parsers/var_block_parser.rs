use crate::{abstract_syntax_tree::AbstractSyntaxNode, parser::ParserResult, Parser};
use nv_lexer::{
    tokens::LexerDeclarationKeyword, LexerKeyword, LexerTokenKind, LexerVarModifierKeyword,
};

use super::var_declaration_parser::VarDeclarationParser;

pub struct VarBlockParser {
    pub ast_block: Vec<AbstractSyntaxNode>,
    pub tokens: Vec<LexerTokenKind>,

    buffer: Vec<LexerTokenKind>,
}

impl Parser for VarBlockParser {
    fn parse(&mut self) -> usize {
        let bound_tokens = self.tokens.clone();
        let mut tokens = bound_tokens.iter().enumerate();

        self.buffer.clear();

        let mut processed_count = 0;

        while let Some((index, token)) = tokens.next() {
            processed_count += 1;
            let sub_tokens = &bound_tokens[index..].to_vec();
            let sub_tokens = sub_tokens.to_vec();

            let result = match token {
                LexerTokenKind::Keyword(LexerKeyword::VarModifierKeyword(
                    LexerVarModifierKeyword::Pub,
                ))
                | LexerTokenKind::Keyword(LexerKeyword::DeclarationKeyword(
                    LexerDeclarationKeyword::Var,
                )) => {
                    let mut parser = VarDeclarationParser::new(sub_tokens.clone());
                    // -1 to avoid double counting the leading token (var or pub)
                    let count = parser.parse() - 1;

                    if let Some(ast_fragment) = parser.ast_fragment.clone() {
                        self.ast_block.push(ast_fragment);
                    }

                    ParserResult {
                        processed_count: count,
                        ast_fragment: parser.ast_fragment,
                    }
                }
                _ => {
                    continue;
                }
            };

            processed_count += result.processed_count;
            if result.processed_count > 0 {
                tokens.nth(result.processed_count - 1);
            }
        }

        processed_count
    }
}

impl VarBlockParser {
    pub fn new(tokens: Vec<LexerTokenKind>) -> Self {
        Self {
            ast_block: vec![],
            tokens,
            buffer: vec![],
        }
    }
}
