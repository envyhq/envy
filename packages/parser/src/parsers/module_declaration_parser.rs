use super::var_block_parser::VarBlockParser;
use crate::{
    abstract_syntax_tree::{AbstractSyntaxNode, DeclarationNode},
    modules::{ModuleDeclarationNode, PartialModuleDeclarationNode},
    Parser,
};
use nv_lexer::{
    tokens::{LexerSymbol, LexerToken},
    LexerTokenKind,
};
use std::sync::Weak;

pub struct ModuleDeclarationParser;

impl Parser<Option<AbstractSyntaxNode>> for ModuleDeclarationParser {
    fn parse(
        tokens: &[LexerToken],
        parent: Weak<AbstractSyntaxNode>,
    ) -> (usize, Option<AbstractSyntaxNode>) {
        let mut buffer = vec![];
        let mut tokens_iter = tokens.iter().enumerate();

        buffer.clear();

        let mut processed_count = 0;

        let mut partial_declaration: PartialModuleDeclarationNode = PartialModuleDeclarationNode {
            identifier: None,
            declarations: vec![],
        };

        while let Some((index, token)) = tokens_iter.next() {
            processed_count += 1;

            let sub_tokens = &tokens[index..];

            match &token.kind {
                LexerTokenKind::Identifier(identifier) => {
                    partial_declaration.identifier = Some(identifier.clone());

                    continue;
                }
                LexerTokenKind::Symbol(LexerSymbol::BlockOpenCurly) => {
                    let (count, parsed_block) = VarBlockParser::parse(sub_tokens, parent.clone());

                    // -1 because we dont want to double count the block open curly
                    let count = count - 1;

                    partial_declaration.declarations = parsed_block
                        .iter()
                        .map(|declaration| declaration.clone().into())
                        .collect();

                    processed_count += count;
                    if count > 0 {
                        tokens_iter.nth(count - 1);
                    }

                    continue;
                }
                LexerTokenKind::Symbol(LexerSymbol::Newline)
                | LexerTokenKind::Symbol(LexerSymbol::BlockCloseCurly) => {
                    break;
                }
                _ => {
                    buffer.push(token);
                    continue;
                }
            };
        }

        let declaration: Result<ModuleDeclarationNode, _> = partial_declaration.clone().try_into();

        let node = declaration
            .ok()
            .map(|d| AbstractSyntaxNode::Declaration(DeclarationNode::ModuleDeclaration(d)));

        (processed_count, node)
    }
}

#[cfg(test)]
mod tests {
    use std::sync::Weak;

    use crate::parsers::module_declaration_parser::{ModuleDeclarationParser, Parser};
    use nv_lexer::{Lexer, ProviderDeclarationLexer, TokenPosition};
    use nv_unit_testing::str_to_graphemes;

    #[test]
    fn parses_module_nodes() {
        // We dont have "provider" keyword here, its handled by the source file parser
        let input = str_to_graphemes("Env: env");

        let start_line = 20;
        let start_column = 5;
        let mut lexer =
            ProviderDeclarationLexer::new(&input, TokenPosition::new(start_line, start_column));
        lexer.lex();

        let parent = Weak::new();
        let (count, ast) = ModuleDeclarationParser::parse(&lexer.tokens, parent);

        assert_eq!(count, 3);

        insta::assert_debug_snapshot!(ast);
    }
}
