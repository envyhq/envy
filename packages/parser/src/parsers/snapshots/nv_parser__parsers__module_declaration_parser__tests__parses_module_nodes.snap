---
source: packages/parser/src/parsers/module_declaration_parser.rs
expression: ast
---
Some(
    Declaration(
        ModuleDeclaration(
            ModuleDeclarationNode {
                identifier: "Env",
                declarations: [],
            },
        ),
    ),
)
